packages/filesystem/src/browser/file-tree/file-tree-model.ts

```diff
@@ -153,9 +153,9 @@ export class FileTreeModel extends CompressedTreeModel implements LocationServic
     async copy(source: URI, target: Readonly<FileStatNode>): Promise<URI> {
         let targetUri = target.uri.resolve(source.path.base);
         try {
-            if (source.path.toString() === target.uri.path.toString()) {
+            if (source.path.toString() === target.uri.path.toString() || source.path.toString() === targetUri.path.toString()) {
```

Comment:
Suggestion: This additional check feels like an unclean solution to the problem. Instead, I would first try to identify potential parent node of the target node and then perform a check. The full method should look something like this:

```ts
async copy(source: URI, target: Readonly<FileStatNode>): Promise<URI> {
  const dirNode = target.fileStat.isFile ? target.parent : target;
  if (!FileStatNode.is(dirNode)) {
    throw new Error('Parent of file has to be a FileStatNode');
  }
  let targetUri = dirNode.uri.resolve(source.path.base);
  try {
    if (source.path.toString() === targetUri.path.toString()) {
      const parent = await this.fileService.resolve(source.parent);
      const name = source.path.name + FileSystemUtils.FILE_NAME_SEPARATOR + 'copy';
      targetUri = FileSystemUtils.generateUniqueResourceURI(source.parent, parent, name, source.path.ext);
    }
    await this.fileService.copy(source, targetUri);
  } catch (e) {
    this.messageService.error(e.message);
  }
  return targetUri;
}
```

Reply:
I applied your suggestion in the new commit. There's one change in how the `targetUri` is computed: we need to append the source base  to the `dirNode` with `uri.resolve` only for directories.

Reply:
Perhaps my understanding of certain file systems is wrong, but isn't the `parent` node of a file node always a directory node? In what case can `dirNode` be a file?

Reply:
I'm sorry, I misspoke. What I meant to say is that when initially computing the targetUri (now line 158 in [file-tree-model.ts](https://github.com/eclipse-theia/theia/pull/10767/files#diff-a25c896256c15be09f9da9d145e14a4f92697ba18ac8612dcd2e1418be759f61R158)), we need to append the source path base to the dirNode only when the target is a _file_ (not a directory, as I initially wrote).

Reply:
Ah, I see. That makes sense 👍